### PR TITLE
[AMORO-2579] Skip the direcotories if can not be deleted when expire snapshots

### LINKED ISSUE #2579: [Bug]: The process of deleting expired snapshots was interrupted by an exception
### What happened?

The process of deleting expired snapshots or files was interrupted by any exception in `TableFileUtil.deleteEmptyDirectory`.
I meet a issue is that: directory is not exist. This issue would be occured when delete a sub-direcotory before the upper empty directory. For example:
```
- data/partition1
- data/partition2
- data/partition2/sub-partition2-1
```
There are three empty folders we need to delete. But we delete the `data/partition2/sub-partition2-1`first, TableFileUtil will also delete the folder `data/partition2`.

### Affects Versions

master/0.6.x

### What engines are you seeing the problem on?

AMS

### How to reproduce

_No response_

### Relevant log output

```shell
2024-02-27 03:32:28,837 ERROR [ASYNC-SnapshotsExpiringExecutor-8] [com.netease.arctic.server.table.executor.SnapshotsExpiringExecutor] [] - unexpected expire error of table catalog.db.xxx(tableId=9510007) 
java.lang.IllegalArgumentException: The target directory is not exist
        at org.apache.iceberg.relocated.com.google.common.base.Preconditions.checkArgument(Preconditions.java:145) ~[iceberg-bundled-guava-1.3.0.jar:?]
        at com.netease.arctic.utils.TableFileUtil.deleteEmptyDirectory(TableFileUtil.java:74) ~[amoro-core-0.6.1-SNAPSHOT.jar:?]
        at com.netease.arctic.server.optimizing.maintainer.IcebergTableMaintainer.lambda$expireSnapshots$1(IcebergTableMaintainer.java:151) ~[amoro-ams-server-0.6.1-SNAPSHOT.jar:?]
        at java.lang.Iterable.forEach(Iterable.java:75) ~[?:1.8.0_392]
        at com.netease.arctic.server.optimizing.maintainer.IcebergTableMaintainer.expireSnapshots(IcebergTableMaintainer.java:150) ~[amoro-ams-server-0.6.1-SNAPSHOT.jar:?]
        at com.netease.arctic.server.optimizing.maintainer.IcebergTableMaintainer.expireSnapshots(IcebergTableMaintainer.java:112) ~[amoro-ams-server-0.6.1-SNAPSHOT.jar:?]
        at com.netease.arctic.server.table.executor.SnapshotsExpiringExecutor.execute(SnapshotsExpiringExecutor.java:60) ~[amoro-ams-server-0.6.1-SNAPSHOT.jar:?]
        at com.netease.arctic.server.table.executor.BaseTableExecutor.executeTask(BaseTableExecutor.java:72) ~[amoro-ams-server-0.6.1-SNAPSHOT.jar:?]
        at com.netease.arctic.server.table.executor.BaseTableExecutor.lambda$scheduleIfNecessary$4(BaseTableExecutor.java:81) ~[amoro-ams-server-0.6.1-SNAPSHOT.jar:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:1.8.0_392]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_392]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) [?:1.8.0_392]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) [?:1.8.0_392]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_392]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_392]
        at java.lang.Thread.run(Thread.java:750) [?:1.8.0_392]
```

### Anything else

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by XBaith**
## Why are the changes needed?

Close #2579.

## Brief change log

- Remove as many empty directories as possible

## How was this patch tested?

- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

- Does this pull request introduce a new feature? (no)
- If yes, how is the feature documented? (not documented)

### CODE REVIEW

File: ams/server/src/main/java/com/netease/arctic/server/optimizing/maintainer/IcebergTableMaintainer.java [hunk 1]
```diff
@@ -183,14 +183,14 @@ private void expireSnapshots(long olderThan, Set<String> exclude) {
         .expireOlderThan(olderThan)
         .deleteWith(
             file -> {
-              try {
+              try (ArcticFileIO io = arcticFileIO()) {
```
Comment (zhoujinsong) [inline ams/server/src/main/java/com/netease/arctic/server/optimizing/maintainer/IcebergTableMaintainer.java:186]: I don't think introducing try-with-resources here is a good approach. The IO objects in the table are not suitable to be closed every time they are used, even though closing them now does nothing.

I understand that IDEA would currently raise a warning here. Perhaps we can try to find some other methods to avoid reporting an error.

File: ams/server/src/main/java/com/netease/arctic/server/optimizing/maintainer/IcebergTableMaintainer.java [hunk 2]
```diff
@@ -203,11 +203,19 @@ private void expireSnapshots(long olderThan, Set<String> exclude) {
...
-    if (arcticFileIO().supportFileSystemOperations()) {
-      parentDirectory.forEach(
-          parent -> TableFileUtil.deleteEmptyDirectory(arcticFileIO(), parent, exclude));
+    try (ArcticFileIO io = arcticFileIO()) {
+      parentDirectory.forEach(parent -> TableFileUtil.deleteEmptyDirectory(io, parent, exclude));
+    } catch (IllegalStateException e) {
+      // Doesn't support directory operations
+      throw e;
+    } catch (Exception ignore) {
+      // Ignore exceptions to remove as many directories as possible
```
Comment (zhoujinsong) [inline ams/server/src/main/java/com/netease/arctic/server/optimizing/maintainer/IcebergTableMaintainer.java:212]: How about logging a warn message here?

File: ams/server/src/main/java/com/netease/arctic/server/optimizing/maintainer/IcebergTableMaintainer.java [hunk 3]
```diff
@@ -203,11 +203,19 @@ private void expireSnapshots(long olderThan, Set<String> exclude) {
...
         .cleanExpiredFiles(true)
         .commit();
-    if (arcticFileIO().supportFileSystemOperations()) {
-      parentDirectory.forEach(
-          parent -> TableFileUtil.deleteEmptyDirectory(arcticFileIO(), parent, exclude));
+    try (ArcticFileIO io = arcticFileIO()) {
+      parentDirectory.forEach(parent -> TableFileUtil.deleteEmptyDirectory(io, parent, exclude));
+    } catch (IllegalStateException e) {
+      // Doesn't support directory operations
+      throw e;
```
Comment (zhoujinsong) [inline ams/server/src/main/java/com/netease/arctic/server/optimizing/maintainer/IcebergTableMaintainer.java:210]: I think we should support FileIO which do not support directory operations and skip the deleting empty directories operations as as we did before rather than throwing an exception.

**Review by zhoujinsong — COMMENTED** [review]
@XBaith I left some comments, PTAL.

File: ams/server/src/main/java/com/netease/arctic/server/optimizing/maintainer/IcebergTableMaintainer.java (hunk 2 above)
Comment (XBaith) [reply ams/server/src/main/java/com/netease/arctic/server/optimizing/maintainer/IcebergTableMaintainer.java]: Make sense!

**Review by zhoujinsong — APPROVED** [review]
LGTM.
